[DESIGN] Effect of selectors on placeholders (#755)
unicode-org/message-format-wg: exploration/selection-declaration.md

# Effect of Selectors on Subsequent Placeholders

Status: **Proposed**

<details>
	<summary>Metadata</summary>
	<dl>
		<dt>Contributors</dt>
		<dd>@aphillips</dd>
		<dt>First proposed</dt>
		<dd>2024-03-27</dd>
		<dt>Pull Requests</dt>
		<dd>#000</dd>
	</dl>
</details>

## Objective

_What is this proposal trying to achieve?_

Define what effect (if any) the _annotation_ of a _selector_ has on subsequent _placeholders_

> that access the same _operand_ value.

Author: Variable is good.

   one {{This case can never happen in an English locale}}
   *   {{I expect this formats num with one decimal place: {$num}}}
   ```

2. As a user, I want to use the least amount of MF special syntax possible.
3. As a user, I don't want to repeat formatting, particularly in large selection matrices.
   ```
   .match {$num1 :integer} {$num2 :number minimumFractionDigits=1}
   0    0    {{You have {$num1 :integer} ({$num2 :number minimumFractionDigits=1}) wildebeest.}}
   0    one  {{You have {$num1 :integer} ({$num2 :number minimumFractionDigits=1}) wildebeest.}}
   0    *    {{You have {$num1 :integer} ({$num2 :number minimumFractionDigits=1}) wildebeest.}}
   one  0    {{ }}
   one  one  {{ }}
   one  *    {{ }}
   // more cases for other locales that use two/few/many
   *    0    {{ }}
   *    one  {{ }}
   *    *    {{ }}
   ```

4. As a user (especially as a translator), I don't want to have to modify
   declarations and selectors to keep them in sync.
   ```
   .input {$num :number minimumFractionDigits=1}
   .match {$num}
   * {{Shouldn't have to modify the selector}}
   ```
   > Note that this is currently provided for by the spec.

### Multiple Selection Conflicts

> If a user writes multiple selectors for the same operand, which one formats the placeholder?

Author: Done.

```
.match {$num :integer} {$num :number minimumFractionDigits=2}
* * {{Which selector formats {$num}?}}

.match {$num :number minimumFractionDigits=2} {$num :integer}
* * {{Which selector formats {$num}?}}

> ```

Author: I added a "Do nothing" option that hopefully illustrates

If both formats are needed in the message (presumably they are or why the selector), 
how does one reference one or the other?

### Selection Is Different From Format
If a user wants to separate selection and formatting, 

> but selectors have a formatting side-effect,

Author: I converted the bottom two sub-section-y things to user stories and reworded.

how do I write a selector that doesn't interfere with the formatting of the _same_ operand?

> (Note that users could do this using a `.local` declaration to split the operand)

Author: Refactored and included an example.

We don't support selection on dates in LDML45, but it's easy to conceptualize.
How can a user write a _selector_ that doesn't mess up a _formatter_?

```
.input {$d :datetime skeleton=yMMMdjm}
.match {$d :datetime month=numeric}
1 {{Today is {$d} in cold cold {$d :datetime month=long}}}
* {{Today is {$d}}}
```

## Requirements

_What properties does the solution have to manifest to enable the use-cases above?_

## Constraints

_What prior decisions and existing conditions limit the possible design?_

## Proposed Design

_Describe the proposed solution. Consider syntax, formatting, errors, registry, tooling, interchange._

## Alternatives Considered

_What other solutions are available?_
_How do they compare against the requirements?_
_What other properties they have?_

### Allow both local and input declarative selectors with immutability

In this alternative, we modify the syntax to allow selectors to 
annotate an input variable (as with `.input`)
or assign a local variable (as with `.local`).
Either annotation is immutable and results in a Duplicate Declaration error

> if it attempts to annotate a variable previously annotated.

Reviewer: ```suggestion
if it attempts to re-declare a previously declared variable.
```

Example:
```
.match {$input :function} $local = {$input :function}
* * {{This annotates {$input} and assigns {$local} a value.}}

.match $local1 = {$input :function} $local2 = {$input :function2}
* * {{This assigns two locals}}

.input {$input :function}
.local $local = {$input :function}
.match {$input :function} {$local :function}

> * * {{This produces two duplicate declaration errors.}}

Reviewer: Why? I only see one declaration for each of the two variables.

Author: `.input {$input}` and `.match {$input}` both "declare" `$input` in this design. Ditto the `.local` and later selector.

```

The ABNF change looks like:
```abnf
selector          = expression / declaration
declaration       = s variable [s] "=" [s] expression
```

**Pros**
- Shorthand is consistent with the rest of the syntax
- Shorthand version works intuitively with minimal input
- Preserves immutability
- Produces an error when users inappropriately annotate some items

**Cons**
- Selectors can't provide additional selection-specific options

>   if the value has already been annotated

Reviewer: ```suggestion
  if the variable name is already in scope
```

- Doesn't allow multiple selection on the same operand, e.g.
  ```
  .input {$date :datetime skeleton=yMdjm}
  .match {$date :datetime field=month} {$date :datetime field=dayOfWeek}
  * * {{This message produces a Duplicate Declaration error

>         even though selection is separate from formatting.}}

Reviewer: As with the previous example, I'm confused. I thought that in this alternative, you need the '=' sign for a declaration, and the `.match` here wouldn't introduce any declarations.

Author: I re-titled this option. In this option, we do not have the `.local`-like feature. These selectors behave exactly like `.input` declarations. A `.input` is a declaration too and doesn't use an `=` sign.

  ```
  However, this design does allow for a local variable to be easily created
  for the purpose of selection.

> #### Allow "Declarative" Selectors with _Mutability_

Author: I re-titled, but I kept the words "mutability" and "immutability" because I think it's a lot clearer than talking about name shadowing. YMMV.

In this alternative, selectors are treated as declaration-selectors.

> That is, an annotation in a selector works like a `.input`.

Reviewer: That doesn't seem right to me. `.input` can only take an argument (free variable), but I think what you're proposing is for selectors to also shadow local variables (bound variables).

Author: Nope. Only the first (well, now second after "do nothing") option provides a local variable binding (which alters the syntax!)

> This permits `.match` selectors to be a shorthand when no declarations exist.

Reviewer: I don't understand what this means.

> It is not an error to overwrite a previous annotation.

Reviewer: ```suggestion
It is not an error to re-declare a variable that is in scope.
```

> Instead the selector's annotation replaces what came before.

Reviewer: ```suggestion
Instead the selector implicitly shadows the name of its variable operand.
```

(This is another place where the distinction between literal and variable operands is important. You can shadow variables, but changing the value of a literal doesn't make sense.)

```
.input {$num :integer}
.match {$num :number minimumFractionDigits=1}
* {{Formats {$num} like 1.0}}
```

**Pros**
- Shorthand version works intuitively with minimal typing.

**Cons**
- Violates immutability that we've established everywhere else

#### Allow "Declarative" Selectors with _Immutability_

In this alternative, selectors are treated as declaration-selectors.
That is, an annotation in a selector works like a `.input`.

> However, it is an error for the selector to try to modify a previous declaration

Reviewer: ```suggestion
However, it is an error for the selector to shadow a previous declaration
```

(just as it is an error for a declaration to try to modify a previous declaration).
This permits `.match` selectors to be a shorthand when no declarations exist.

> It is also an error for a selector to modify a previous selector.

Reviewer: ```suggestion
It is also an error for a selector to shadow a previous selector.
```

(Again, this needs some qualifiers... writing `.match {-1 :integer} {-1 :number}` shouldn't be an error.)

Author: The example you gave doesn't modify a previous selector: the literal `-1` in the first is a separate literal from that in the second. "Shadow" doesn't really say what is going on?

Reviewer: If a selector implicitly introduces a name, then what you're referring to as "modifying" is actually "shadowing". If you write, in this scenario:

```
.match {$x :integer} {$x :number} 
```

then the second `$x` doesn't modify the first one. Rather, it introduces a new name `$x`, so that in the patterns in the variants, `$x` now has a different meaning. That's what I mean by name shadowing.

> This implies that multiple selecton on the same operand is pointless.

Reviewer: Not so much "pointless" as "always emitting an error"?

Author: No, one can do this just fine:
```
.input {$foo :number}
.match {$foo} {$foo}
one one {{why select twice, when the second selector will rank the same as the first?}}
```

```
.match {$num :integer}
* {{Formats {$num} as integer}}

.input {$num :integer}
.match {$num :number maximumFractionDigits=0}
* {{This message produces a Duplicate Declaration error}}

.input {$num :integer} {$num :number}
* * {{This message produces a Duplicate Declaration error}}
```

**Pros**
- Shorthand version works intuitively with minimal typing
- Preserves immutability
- Produces an error when users inappropriately annotate some items

**Cons**
- Selectors can't provide additional selection-specific options
  if the value has already been annotated
- Doesn't allow multiple selection on the same operand, e.g.
  ```
  .input {$date :datetime skeleton=yMdjm}
  .match {$date :datetime field=month} {$date :datetime field=dayOfWeek}
  * * {{This message produces a Duplicate Declaration error
        even though selection is separate from formatting.}}
  ```

> 

Reviewer 2: Let's add my proposal from https://github.com/unicode-org/message-format-wg/issues/736#issuecomment-2016408105:

````suggestion

#### Match on variables instead of expressions

In this alternative, the `.match` syntax is simplified
to work on variable references rather than expressions:

```
.input {$count :number}
.match $count
one {{You have {$count} apple}}
* {{You have {$count} apples}}

.local $empty = {$theList :isEmpty}
.match $empty
true {{You bought nothing}}
* {{You bought {$theList}!}}
```

The ABNF change would look like:
```diff
 match-statement   = match 1*([s] selector)
-selector          = expression
+selector          = variable
```

**Pros**
- Overall the syntax is simplified.
- Preserves immutability.

**Cons**
- A separate declaration is required for each selector.

````

#### Provide a `#`-like Feature

(Copy-pasta adapted from @eemeli's proposal in #736)

Make the `.match` expression also act as implicit declarations accessed by index position:

```
.match {$count :number}
one {{You have {$0} apple}}
* {{You have {$0} apples}}
```

Assigning values to `$0`, `$1`, ... would not conflict with any input values, 
as numbers are invalid `name-start` characters.
That's by design so that we encourage at least _some_ name for each variable; 
here that's effectively provided by the `.match` expressions.

ABNF would be modified:
```diff
-variable = "$" name
+variable = "$" (name / %x30-39)
```

...with accompanying spec language making numeric variables resolve to the `.match` selectors in placeholders, 
and a data model error otherwise.

**Pros**
- More ergonomic for most `.input` cases
- Enables representation of many messages without any declarations

**Cons**

> - Confusing that the operand name can't be used in the pattern?

Reviewer: Why does the operand name have to be disallowed from appearing in the pattern?

Author: Perhaps not worded correctly. You can use the operand name in the pattern, but it's annotation won't/might not be applied.

```
.match {$count :integer}
* {{You have {$count} apples might print 2 or 2.0 with input count=|2.0|}}
```

Reviewer: Right, so maybe something like: "The meaning of the operand name in the context of the pattern might not be what the user expects"?

  Removes some self-documentation from the pattern.
- Requires the pattern to change if the selectors are modified.

> - Limits number of referenceable selectors to 10 (in the current form)

Reviewer: I think one more alternative is missing, which is to do nothing: require a declaration when you want a declaration instead of introducing a third (maybe third and fourth?) declaration form, and maintain the invariant that the meaning of a variable name is the same regardless of where it appears.

Author: Will add

   one {{This case can never happen in an English locale}}
   *   {{I expect this formats num with one decimal place: {$num}}}
   ```

2. As a user, I want to use the least amount of MF special syntax possible.
3. As a user, I don't want to repeat formatting, particularly in large selection matrices.
   ```
   .match {$num1 :integer} {$num2 :number minimumFractionDigits=1}
   0    0    {{You have {$num1 :integer} ({$num2 :number minimumFractionDigits=1}) wildebeest.}}
   0    one  {{You have {$num1 :integer} ({$num2 :number minimumFractionDigits=1}) wildebeest.}}
   0    *    {{You have {$num1 :integer} ({$num2 :number minimumFractionDigits=1}) wildebeest.}}
   one  0    {{ }}
   one  one  {{ }}
   one  *    {{ }}
   // more cases for other locales that use two/few/many
   *    0    {{ }}
   *    one  {{ }}
   *    *    {{ }}
   ```

4. As a user (especially as a translator), I don't want to have to modify
   declarations and selectors to keep them in sync.
   ```
   .input {$num :number minimumFractionDigits=1}
   .match {$num}
   * {{Shouldn't have to modify the selector}}
   ```
   > Note that this is currently provided for by the spec.

> 5. As a user, I want to write multiple selectors for the same operand.

Reviewer 3: ```suggestion
5. As a user, I don't want to write multiple selectors for the same operand.
```

Author: @gibson042 I don't understand the change proposed? This use case is illustrating a user who _does_ want to select on the same variable using different criteria.
